### ISSUE TITLE
BN_new() gives SEGFAULT

### ISSUE THREAD

**Issue by mayank0403** [intel/intel-sgx-ssl#33]
I have been porting some c++ code to SGX. The ported codebase is huge, so I am refraining from putting the code here.
The issue is that I get a SEGFAULT at a later point in my code (say at time t) if I have ever used `BIGNUM* bne = BN_new();` in my code at time t-, even though I never use this variable `bne` in my code later.
Using SGX-GDB points to the source of segfault at some other line, but removing `bne` declaration and initialization from my code makes everything run fine. I feel, the source of the memory corruption has something to do with BIGNUM only.

TL;DR: Just commenting out `BIGNUM* bne = BN_new();` makes the code run fine, even though `bne` is never used in the future.
I did not link `libsgx_tsetjmp` in my Makefile. Could this be the culprit?
Any help is appreciated. 
Thanks.

**Comment by akryeem-INTC** [comment]
Could you give some more details? SGX SDK version, SGXSSL branch/tag, linux distro. Maybe the enclave is out of memory or has reached max EPC usage? Could you replace the BN_new() with some other ctx allocation (RSA_new() for example, or simply allocate some extra memory).

Also sharing some code for reproducing the issue would be very helpful.

**Comment by mayank0403** [comment]
Here are some details as you asked:
1. Using RSA_new() also causes a SEGFAULT.
2. If I allocate 10000 bytes of memory using malloc, then it runs fine.
3. Here is some code which is run inside the enclave: (`test_main_ckt()` does not use any ssl functionality. At the very crude level, it is based on the [Relic Toolkit](https://github.com/relic-toolkit/relic) which does not use openssl). 
`void sgx_printf(int pnum, const char *fmt, ...)
{
	party_num = pnum;
	struct cmd_pack* pack;
	struct cmd_pack* packsh;
	
	if(pnum == 0){ // ALICE
		pack = cmd_parse("1 4567");
		packsh = cmd_parse("1 4567 120");
	}
	else{ // BOB
		pack = cmd_parse("2 4567");
		packsh = cmd_parse("2 4567 120");

	}
	//RSA* rs = RSA_new();
	//BIGNUM* bne = BN_new();
	int* extra_memory = (int*)malloc(10000);
	if(extra_memory == NULL){
		return;
	}	

	test_main_ckt(pack->argc, pack->argv);
}
`
If I uncomment any of the lines with BN_new() or RSA_new(), I get SEGFAULT.
4. SGX SDK version 2.2.1 (x64)
    OS: Ubuntu 16.04
    SGX-SSL commit a546f96eecfafda383f747507025ab4fa874e3e8 

Let me know if any other information is required. 
Since, gdb points to the SEGFAULT at a line inside the Relic code, I doubt if a stack dump of that would help.
I can still post it if it is required.

Thanks.

**Comment by mayank0403** [comment]
Here are the link flags.
```
Enclave_C_Flags += $(Enclave_Include_Paths)
Enclave_Cpp_Flags := $(Enclave_C_Flags) -std=c++11 -nostdinc++ -msse4.2 -maes -march=native -fomit-frame-pointer -mpclmul -O3

Enclave_Link_Flags := $(SGX_COMMON_CFLAGS) -Wl,--no-undefined -nostdlib -nodefaultlibs -nostartfiles -L$(GMP_SGX) -lsgx_tgmp -L$(SGX_LIBRARY_PATH) -L$(SSL_SGX)/lib64 -lsgx_tsgxssl -L/usr/local/lib \
	-Wl,--whole-archive -l$(Trts_Library_Name) -Wl,--no-whole-archive \
	-Wl,--start-group -lsgx_tstdc -lsgx_tstdcxx -lsgx_tcxx -l$(Crypto_Library_Name) -l$(Service_Library_Name) -L$(SSL_SGX)/lib64 -lsgx_tsgxssl_crypto -lsgx_tswitchless -lsgx_tprotected_fs -Wl,--end-group \
	-Wl,-Bstatic -Wl,-Bsymbolic -Wl,--no-undefined \
	-Wl,-pie,-eenclave_entry -Wl,--export-dynamic  \
	-Wl,--defsym,__ImageBase=0 -Wl,--gc-sections   \
	-Wl,--version-script=Enclave/Enclave.lds
```

**Comment by akryeem-INTC** [comment]
Looking in the `BN_new()` API, by default what is does at end is a simple `malloc `and `memset 0` to allocated memory. You may try replacing the `BN_new()` API with it's internal code and see why it has different behavior than the direct `malloc` you called. I can't reproduce this issue, giving a reproduce steps would be appreciated.

**Comment by akryeem-INTC** [comment]
Another direction I can think about is linker optimizations when loading OpenSSL data into the enclave. If `BN_new()` is the first and only call to OpenSSL code, hence it prevented optimizations and caused loading all required OpenSSL symbols into your enclave (for example OpenSSL error strings) which was optimized before. Try using other OpenSSL API for example `volatile BIGNUM *a = NULL; BN_free(a);` and see what happens. You can also check the required memory reported by sgx_sign tool when building the enclave, and see if it's different before and after adding the `BN_new()`.

**Comment by akryeem-INTC** [comment]
Closing, reopen in case it still relevant.
